### DISCUSSION TITLE
Use remote ref as protected base?

### DISCUSSION THREAD

**Discussion by arxanas** [gitext-rs/git-stack#93] [Ideas]
I wanted to use `origin/master` as my base branch. I tried

```
$ git stack --protect origin/master
$ git stack
could not find a protected branch to use as a base
```

It only started working once I created a local `master` branch and used that as the base. I don't normally keep a local `master` branch, so would it be possible to use a different ref as the base?

**Comment by epage** [comment]
Currently, I only operate on local branches.  I do have https://github.com/epage/git-stack/issues/12 which this is good input for.

**Comment by Gelio** [comment]
Personally, I would also love to see `git stack` work on remote branches directly. This limitation is the only reason I have to keep a local `master` branch that often gets stale. It would be awesome if it was able to work off `remote/master` directly, so I wouldn't be bothered with keeping a local `master` branch.

Moreover, this would improve the interop with https://github.com/arxanas/git-branchless. If I do `git stack --pull`, `git-branchless` will not know that it can hide the old commits (pre-rebase) and show them in `git sl`. I am using `git move -b HEAD -d origin/master` from `git-branchless` as a substitute for `git stack --pull`.
Because I am not using `git stack --pull`, this keeps my local `master` branch stale. Of course, that's not a direct issue for `git stack`, because that is related to the interoperability between the 2 tools, but not having to rely on a local `master` branch would make using both tools easier

**Reply by arxanas** [reply]
@epage does `git-stack` invoke the appropriate Git hooks after it does work? That should make it work well with `git-branchless` (or potentially other tools), but looking through https://github.com/epage/git-stack/blob/203a06e87a9989bcf43e3bcc7b6678acf6fe86a0/src/git/repo.rs, I didn't see anything like that.

It looks like the relevant hooks might be:
- `post-commit`
- `post-rewrite`
- `post-merge`
- `reference-transaction`

(from here https://github.com/arxanas/git-branchless/blob/496ad073d28382e967fc4e27eeeab26d6dc4c06f/src/commands/init.rs#L17-L60)

**Reply by epage** [reply]
git-stack does not invoke any hooks, I had assumed libgit2 would do it for me.

The bit of research that'll need to be done is under what condition a hook is supposed to be called.

**Reply by Gelio** [reply]
I have created https://github.com/epage/git-stack/issues/192 to track calling hooks by `git-stack`, focusing mostly on the interop with `git-branchless`.

**Reply by epage** [reply]
@Gelio and @arxanas  I've started laying groundwork for this but something I was considering is when the pull remote has a lot of branches.  For example, `gecko-dev` seems to have ~600 remote branches.
- We could enumerate all of them and treat them as protected branches.  That would bloat the UI and probably slow things down as we try to determine which feature branch belongs to which protected branch.  
- We could hide protected branches that have no commits but there are valid cases for that (e.g. when you've wrapped up all feature branches, your base disappears) and there is still the potential for performance issues
- We could add a `ignored-branch` setting, like `protected-branch`, but we've been aiming for working out of the gate as much as possible but this could force people to configure `git-stack` just to use it

Thoughts on these trade offs or on other approaches?

**Reply by epage** [reply]
`main` now has some level of remote branch tracking.  I've created #213 for handling the "no local copy" case.
